trio/_core/_run.py

```diff
@@ -780,28 +820,57 @@ def cancel_called(self) -> bool:
 ################################################################
 
 
+class TaskStatus(Protocol[StatusT_contra]):
+    """The interface provided by :meth:`Nursery.start()` to the spawned task.
+
+    This is provided via the ``task_status`` keyword-only parameter.
+    """
+
+    @overload
+    def started(self: TaskStatus[None]) -> None:
+        ...
+
+    @overload
+    def started(self, value: StatusT_contra) -> None:
+        ...
+
+    def started(self, value: StatusT_contra | None = None) -> None:
+        """Tasks call this method to indicate that they have initialized.
+
+        See `nursery.start() <trio.Nursery.start>` for more information.
+        """
+
+
 # This code needs to be read alongside the code from Nursery.start to make
 # sense.
 @attr.s(eq=False, hash=False, repr=False)
-class TaskStatus(metaclass=Final):
-    _old_nursery = attr.ib()
-    _new_nursery = attr.ib()
-    _called_started = attr.ib(default=False)
-    _value = attr.ib(default=None)
+class _TaskStatus(TaskStatus[StatusT]):
+    _old_nursery: Nursery = attr.ib()
+    _new_nursery: Nursery = attr.ib()
+    # NoStatus is a sentinel.
+    _value: StatusT | type[_NoStatus] = attr.ib(default=_NoStatus)
 
-    def __repr__(self):
+    def __repr__(self) -> str:
         return f"<Task status object at {id(self):#x}>"
 
-    def started(self, value=None):
-        if self._called_started:
+    @overload
+    def started(self: _TaskStatus[None]) -> None:
+        ...
+
+    @overload
+    def started(self: _TaskStatus[StatusT], value: StatusT) -> None:
+        ...
+
+    def started(self, value: StatusT | None = None) -> None:
```

Comment:
Technically the `| None` is unnecessary. I don't know if mypy supports that though (I remember seeing a still-open issue that looked roughly like this.)

Reply:
It is absolutely necessary in new versions of mypy. That is a legacy feature that I think they are trying very hard to depreciate.

Reply:
Nah I'm not talking about strict optional: I'm saying that that matches the overloads (cause StatusT would have to be None in order for that default to work)

Reply:
Unfortunately no, Mypy still complains.